[xutility] Modernize _Ptr_meow_cat to use variable templates

miscco opened the pull request:
This is a work in progress for #860 that I started yesterday. 

With @StephanTLavavej awesome work on `_Equal_memcmp_is_safe` it went quite well. 

* I use an enum to differentiate between the different categories. Is that ok or should I use some other means of discriminating between the categories.

* There is some unfortunate code duplication because enable_if does not work on specializations of variable templates. (It complains about an unused template argument). Is that ok or has somebody a better idea?

* Names are hard

- [x] That said there is some bug I need to hunt but I wanted to get the overal feedback on the approach.

- [x] The `unwrap_enum_t` part is terrible I will most likelybuild a helper template for that

miscco commented:
@CaseyCarter This passes without ranges support. The problem is again that `contiguous_iterator<volatile int*>` returns false, which breaks all the tests in C++20 mode.

I seem to remember that you already filed an issue regarding some bug in the ternary operator that leads to common_with to be false but was unable to find it. Has that been resolved?

miscco commented:
I also found that `iter_value_t` breaks everything here. I would have thought I could use it instead of `remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter1>>>` but ...

miscco commented:
I have trouble understanding the test errors, as those specializations they complain do work fine in the test.

Can somebody that understands meta programming help?

miscco commented:
So it seems that the two overloads for *same type copy* and *pointer copy* were indeed ambiguous if `_Source == _Dest`.

I added some SFINAE or that to *same type copy* specializations.

There is still a question why MSVC accepts this code when then other major compilers do.  

I will try to condense this down a bit and create a ticket

miscco commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,105 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-    static constexpr bool _Trivially_copyable = true;
-};
+// Matching integer type are eligible for memcopy
+template <class _Source, class _Dest, _Memcopy_cat _Cat,
+    bool = _Memcopy_matching_integral<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
+_INLINE_VAR constexpr bool _Can_memcopy_elements = true;
```
I think one could improve this by reusing the result from `_Memcopy_matching_integral`, aka
```suggestion
template <class _Source, class _Dest, _Memcopy_cat _Cat,
    bool _Result = _Memcopy_matching_integral<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
_INLINE_VAR constexpr bool _Can_memcopy_elements = _Result;
```

This would make the `// Arbitrary types are not eligible` specialization obsolete.

That said it would be harder to find the default case

miscco commented:
@Casey can I get some direction on how to fix the ostreambuf mess?

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-struct _Ptr_cat_helper<_Elem, _Elem> { // determines _Ptr_cat's result when the types are the same
-    static constexpr bool _Really_trivial     = is_trivial_v<_Elem>;
-    static constexpr bool _Trivially_copyable = is_trivially_copyable_v<_Elem>;
+// _Can_memcopy_elements<_Ty1, _Ty2> reports whether `_Ty1` can be memcopy'd into  _Ty2`.
```
Every occurrence of `memcopy` should be `memcpy`. Vowels were expensive back in the 1970s.

miscco replied:
What we have 2020 now.
...
Looks around what is going on in the world. 
...
I think you meant 1960s...

StephanTLavavej commented on stl/inc/memory:
```diff
@@ -76,8 +76,10 @@ _NoThrowFwdIt uninitialized_copy_n(const _InIt _First, const _Diff _Count_raw, _
-                                 bool_constant<_Ptr_copy_cat<decltype(_UFirst), decltype(_UDest)>::_Really_trivial>{}));
+        _Seek_wrapped(_Dest,
+            _Uninitialized_copy_n_unchecked2(_UFirst, _Count, _UDest,
+                bool_constant<
+                    _Memcopy_is_safe<decltype(_UFirst), decltype(_UDest), _Memcopy_cat::_Copy_uninitialized>>{}));
```
Style: The wrapping is becoming horrible, so this `bool_constant` should be extracted with `using`. (We've often said `using _Tag = BLAH` for this purpose.)

StephanTLavavej commented on stl/inc/xmemory:
```diff
@@ -1626,8 +1628,9 @@ _NoThrowFwdIt uninitialized_copy(const _InIt _First, const _InIt _Last, _NoThrow
-                             bool_constant<_Ptr_copy_cat<decltype(_UFirst), decltype(_UDest)>::_Really_trivial>{}));
+    _Seek_wrapped(_Dest,
+        _Uninitialized_copy_unchecked(_UFirst, _ULast, _UDest,
+            bool_constant<_Memcopy_is_safe<decltype(_UFirst), decltype(_UDest), _Memcopy_cat::_Copy_uninitialized>>{}));
```
Same suggestion for `using` here, to make the `_Seek_wrapped` call significantly easier to read.

StephanTLavavej commented on stl/inc/xmemory:
```diff
@@ -1689,10 +1693,11 @@ _Alloc_ptr_t<_Alloc> _Uninitialized_move(
-                       && _Uses_default_construct<_Alloc, _Ptrval, decltype(_STD move(*_UFirst))>::value > {}));
+    _Seek_wrapped(_Dest,
+        _Uninitialized_move_al_unchecked(_UFirst, _ULast, _UDest, _Al,
+            bool_constant < _Memcopy_is_safe<decltype(_UFirst), _Ptrval,
```
Another statement that could really benefit from `using` (and wow, the formatting was already terrible).

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-struct _Ptr_cat_helper<_Elem, _Elem> { // determines _Ptr_cat's result when the types are the same
-    static constexpr bool _Really_trivial     = is_trivial_v<_Elem>;
-    static constexpr bool _Trivially_copyable = is_trivially_copyable_v<_Elem>;
+// _Can_memcopy_elements<_Ty1, _Ty2> reports whether `_Ty1` can be memcopy'd into  _Ty2`.
```
Two spaces and missing backtick before `_Ty2`.

miscco replied:
Fixed

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-    static constexpr bool _Trivially_copyable = is_trivially_copyable_v<_Elem>;
+// _Can_memcopy_elements<_Ty1, _Ty2> reports whether `_Ty1` can be memcopy'd into  _Ty2`.
+// Here, _Ty1 and _Ty2 aren't top-level const, because we remove_const_t before using _Can_memcopy_elements.
+// There are different specialization wrt move / copy and whether the destination is uninitialized memory
```
"specialization" is a Word Of Power, so I find it confusing here. Overall, this sentence doesn't seem to be explaining much beyond what the names of the enumerators already say.

miscco replied:
Removed that sentence

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-};
+// Integral types are eligible for memcopy in very specific cases.
+// * Enumerations are treated as their underlying integral type
+// * They must be the same size. (`int == long` is eligible; `int == long long` isn't.)
```
The `==` here is copy-pasted.

miscco replied:
Done

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-};
+// Pointer elements are eligible for memcopy when they point to the same type, ignoring cv-qualification.
+// This handles pointers to object types, pointers to void, and pointers to function types.
+// Performance note: This doesn't attempt to handle `object* == void*`.
```
Same `==` copy-pasting.

miscco replied:
Done

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+    is_trivially_assignable_v<_Dest&, _Source>;
+
+// _Memcopy_is_safe<_Iter1, _Iter2> reports whether we can activate the memcopy optimization for arbitrary iterators.
+// It ignores top-level constness on the iterators and on the elements.
```
This is not quite correct - unlike the `equal` machinery, this cares about assigning to const elements. In general, when copy-paste-modifying code and comments, everything needs to be audited to make sure that it still applies exactly.

miscco replied:
I removed the `remove_const` from both the iterators as well as the elements of dest

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+    _Memcopy_is_safe_helper<remove_const_t<_Iter1>, remove_const_t<_Iter2>, _Cat>;
+
+template <class _Iter1, class _Iter2, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Memcopy_is_safe<move_iterator<_Iter1>, _Iter2, _Cat> =
```
I believe that this fails to handle a top-level `const move_iterator`.

miscco replied:
urgh I am checking whether I find some solution

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+        remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter2>>>, _Cat>>>;
+
+template <class _Iter1, class _Iter2, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Memcopy_is_safe =
```
Usage now looks like `_Memcpy_is_safe<MEOW, _Memcpy_cat::_Copy>` which ends up saying "copy" (or close enough) three times. Should we have alias templates for the four different categories?

miscco replied:
I am torn between "Thanks for this totally sensible suggestion" and "Why did I not think about that?"

I will go over this hopefully later this evening and will adapt the naming scheme.

miscco replied:
Added aliases

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+template <class _Source, class _Dest, _Memcopy_cat>
+_INLINE_VAR constexpr bool _Can_assign_elements = is_trivially_assignable_v<_Dest&, _Source&>;
+
+// Different types are eligible for memmove when they are trivially copy assignable.
```
I am confused by this comment - it is referring to `memmove` (which handles overlapping) but the arguments to the type trait indicate that the source is being treated as an rvalue (i.e. C++ moving). I think that the comment saying "trivially copy assignable" should be "trivially move assignable"?

miscco replied:
"trivially move assignable" was the correct phrase

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4341,8 +4379,9 @@ _BidIt2 copy_backward(_BidIt1 _First, _BidIt1 _Last, _BidIt2 _Dest) { // copy [_
-                             bool_constant<_Ptr_copy_cat<decltype(_UFirst), decltype(_UDest)>::_Trivially_copyable>{}));
+    _Seek_wrapped(
+        _Dest, _Copy_backward_unchecked(_UFirst, _ULast, _UDest,
+                   bool_constant<_Memcopy_is_safe<decltype(_UFirst), decltype(_UDest), _Memcopy_cat::_Copy>>{}));
```
Being 3 lines now, this could benefit from `using`.

miscco replied:
Done

StephanTLavavej commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -186,6 +186,17 @@ void ptr_cat_test_cases() {
+    test_ptr_cat<2, trivially_copyable_struct**, trivially_copyable_struct**>();
+    test_ptr_cat<2, custom_copy_struct**, custom_copy_struct**>();
+
+    // Pointer to pointer of different types not
```
This comment needs a bit of grammar work. (Test code comments don't have to be perfectly phrased - they can be sentence fragments, but this one is too fragmented.)

miscco replied:
Done

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+_INLINE_VAR constexpr bool _Memcopy_is_safe_helper = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>,
+    bool_constant<
+        _Can_assign_elements<remove_reference_t<_Iter_ref_t<_Iter1>>, remove_reference_t<_Iter_ref_t<_Iter2>>, _Cat>>,
+    bool_constant<_Can_memcopy_elements<remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter1>>>,
+        remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter2>>>, _Cat>>>;
```
I guess it was pre-existing (and prevented from mattering by the "is trivially assignable" type trait check), but should we be removing top-level volatility here now?

miscco commented:
I addressed most of the stylistic comments and will have a look on how to fix those failures

miscco commented:
This is driving me crazy. Why it is trying to do `sizeof(void)` in  `_Can_memcpy_integrals` when `is_integral<void>` should already have returned false?

AdamBucior commented:
> This is driving me crazy. Why it is trying to do `sizeof(void)` in `_Can_memcpy_integrals` when `is_integral<void>` should already have returned false?

`std::conjunction` short circuiting doesn't work that way. Template parameters need to be evaluated before template is instantiated. You need:
```
template <class _Source, class _Dest>
struct _Find_some_name_for_this_struct
{
    static constexpr bool value = sizeof(_Source) == sizeof(_Dest) && is_same_v<bool, _Source> == is_same_v<bool, _Dest>;
};
```
and then:
```
constexpr bool _Can_memcpy_integrals = conjunction_v<is_integral<_Source>, is_integral<_Dest>,
    _Find_some_name_for_this_struct<_Source, _Dest>>;
```

miscco commented:
> 
> 
> > This is driving me crazy. Why it is trying to do `sizeof(void)` in `_Can_memcpy_integrals` when `is_integral<void>` should already have returned false?
> 
> `std::conjunction` short circuiting doesn't work that way. Template parameters need to be evaluated before template is instantiated. You need:
> 
> ```
> template <class _Source, class _Dest>
> struct _Find_some_name_for_this_struct
> {
>     static constexpr bool value = sizeof(_Source) == sizeof(_Dest) && is_same_v<bool, _Source> == is_same_v<bool, _Dest>;
> };
> ```
> 
> and then:
> 
> ```
> constexpr bool _Can_memcpy_integrals = conjunction_v<is_integral<_Source>, is_integral<_Dest>,
>     _Find_some_name_for_this_struct<_Source, _Dest>>;
> ```

Argh damn it, I am always forgetting that...

miscco commented:
I am super unhappy that I had to essentially roll back the most useful changes to go back to that pointer only specialization. 

However, it seems like there are some deficiencies wrt `iter_reference_t` and `iter_value_t`

miscco commented:
I think this should be ready for the next round of reviews

mnatsuhara commented on stl/inc/xutility:
```diff
@@ -4043,69 +4043,111 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-    static constexpr bool _Trivially_copyable = _Really_trivial;
-};
+// _Iterators_are_contiguous<_Iter1, _Iter2> reports whether both iterators are known to be contiguous.
+// (Without concepts, this detection is limited, which will limit when we can activate certain optimization.)
```
Being a bit picky here, but should this be `certain optimizations` or `a certain optimization`?  I feel like an article may be missing here...

miscco replied:
Reading it aloud you are correct. Maybe

```suggestion
// (Without concepts, this detection is limited, which will limit when we can activate the given optimization.)
```

miscco replied:
Done

mnatsuhara commented on stl/inc/xmemory:
```diff
@@ -1689,10 +1689,10 @@ _Alloc_ptr_t<_Alloc> _Uninitialized_move(
-                       && _Uses_default_construct<_Alloc, _Ptrval, decltype(_STD move(*_UFirst))>::value > {}));
+    _Seek_wrapped(_Dest,
+        _Uninitialized_move_al_unchecked(_UFirst, _ULast, _UDest, _Al,
+            bool_constant < _Can_memmove_uninitialized<decltype(_UFirst),
+                    _Ptrval> && _Uses_default_construct<_Alloc, _Ptrval, decltype(_STD move(*_UFirst))>::value > {}));
```
It seems like the spacing around `bool_constant` is different here than in other occurrences (specifically before and after the `<` and `>`), and it also looks like it is preexisting in the code before changes.  Is there a reason for this?

miscco replied:
I believe this is clang format getting confused by the `_Uses_default_construct<_Alloc, _Ptrval, decltype(_STD move(*_UFirst))>::value > {}` afterwards

I was not able to create a simple reproducer though

StephanTLavavej commented:
There's a merge conflict in `xutility` now.

miscco commented:
Given that this PR has been a bit stale for some time and I got some strange problems with messed up formatting after merging master in, I have sqaush rebased the whole thing

StephanTLavavej commented:
There's a merge conflict in `<memory>` now.

AdamBucior commented:
The names of variable templates are misleading and inconsistent. `_Can_memcpy` should be `_Copy_memmove_is_safe`, `_Can_memcpy_uninitialized` should be `_Uninitialized_copy_memcpy_is_safe`, `_Can_memmove` should be `_Move_memmove_is_safe` and `_Can_memmove_uninitialized` should be `_Uninitialized_move_memcpy_is_safe`.

AdamBucior commented on stl/inc/xutility:
```diff
@@ -4299,69 +4299,111 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for copy
+// Types are eligible for memcpy / memmove when they are trivially copyable.
```
What is the reasoning behind this? Types should be eligible for using `memcpy` / `memmove` when they are trivially copy constructible / move constructible / copy assignable / move assignable based on the operation. `copy` simply invokes copy assignment operator for every element so it should use `memmove` whenever the elements are trivially copy assignable. Similarly `move` simply invokes move assignment operator, `uninitialized_copy` simply invokes copy constructor and `uninitialized_move` simply invokes move constructor. `move` should not care about whether type is trivially copy assignable so `is_trivially_copyable` is too strict. Same with uninitialized versions which for some reason require type to be trivially default constructible (so types like `string_view` which is trivially copyable but not trivially default constructible won't be optimized).

miscco replied:
So this is just reformulating the existing code to use variable templates. I would rather not make functional changes within this PR if not needed.

I fully agree that we should have a more fine grained approach and this PR tries to build a flexible system so that we can more easily compose the requirements. This is essentially done by splitting it up into three independent parts

1. Are the iterators contiguous?
2. Does the type fulfill the necessary requirements?
3. Does the operation fulfill the necessary requirements?

AdamBucior replied:
I understand. Still I would like a maintainer to answer my question.

StephanTLavavej replied:
I believe the reasoning here is simply that this code is very old - the core of the optimization (detecting types that are safe to `memcpy/memmove`) was present in the codebase when I joined in 2007. My memories may be imperfect - this was one decade and two source control systems ago - but I seem to recall the original optimization being limited to only arithmetic and pointer types. I believe we then extended it to `is_pod` (which was extremely buggy at the time), and later that was reworked when `is_pod` became deprecated.

So the answer to the mystery of why we aren't using fine-grained type traits here is simply that they didn't exist back then. Also, we were confused by the fine-grained type traits for a while (there was a bug where `is_trivially_copy_constructible` was implemented with `is_trivially_copyable` - I missed all sorts of amazing bugs in the old days).

I think that separate issues/PRs to refine these optimizations to query the exact type traits needed would make a lot of sense. Thanks for bringing this up!

(I am having some success at clearing out my GitHub notifications but always feel free to ping us on Discord too if questions come up.)

miscco commented:
> The names of variable templates are misleading and inconsistent. `_Can_memcpy` should be `_Copy_memmove_is_safe`, `_Can_memcpy_uninitialized` should be `_Uninitialized_copy_memcpy_is_safe`, `_Can_memmove` should be `_Move_memmove_is_safe` and `_Can_memmove_uninitialized` should be `_Uninitialized_move_memcpy_is_safe`.

I have to disagree here:

1. `can_meow` vs `meow_is_safe`
This is a purely aesthetic choice and I prefer the former. Also we use `can_meow`(124 times)  in the code base way more often than `meow_is_safe` (28 times) and nearly all of those `meow_is_safe` occurences come from only two templates `_Equal_memcmp_is_safe` and `_Equal_memset_is_safe`

2. `uninitialized_meow` vs `meow_uninitialized`

Here `meow_uninitialized` is a refinement of `meow`. So if I read `meow` and `meow_uninitialized` it actually aligns clearly. This is not the case for `uninitialized_meow` and `meow`. Also note that `meow` is usually the important aspect and should come front.

This is essentially the reason why in radio communication you open a call with `Bob for Alice` and not `Alice to Bob`. the most relevant information should come first.

3. `copy_memcpy` vs `move_memcpy`
Here I agree that we should have a separation between `copy` and `move`, but not with the additional duplication of `copy_memcpy`. This seems a lot longer than it needs to be. 

That said it is currently not the case and I would prefer to stay with the current functionality as is and add improvements once the pattern is established.

AdamBucior commented:
My main point is that `_Can_memcpy` guards usage of `memmove` in `copy` and `_Can_memmove_uninitialized` guards usage of `memcpy` in `uninitialized_move`. This is very misleading. If you don't like my proposed names you should at least change them to `_Can_memmove_in_copy`, `_Can_memmove_in_move`, `_Can_memcpy_in_uninitialized_copy`, `_Can_memcpy_in_uninitialized_move` (though I still prefer `_Algorithm_mem*_is_safe`)

miscco commented:
> My main point is that `_Can_memcpy` guards usage of `memmove` in `copy` and `_Can_memmove_uninitialized` guards usage of `memcpy` in `uninitialized_move`. This is very misleading. If you don't like my proposed names you should at least change them to `_Can_memmove_in_copy`, `_Can_memmove_in_move`, `_Can_memcpy_in_uninitialized_copy`, `_Can_memcpy_in_uninitialized_move` (though I still prefer `_Algorithm_mem*_is_safe`)

Yeah, I was endlessly confused by that a lot too. I think the problem is that `memmove` and `memcpy` are essentially the same implementation, which leads to those problems. Not that in `copy` we call `_Copy_memmove` which is the root of the naming problem.

StephanTLavavej commented:
> If you don't like my proposed names you should at least change them to `_Can_memmove_in_copy`, `_Can_memmove_in_move`, `_Can_memcpy_in_uninitialized_copy`, `_Can_memcpy_in_uninitialized_move` (though I still prefer `_Algorithm_mem*_is_safe`)

At first glance, I like the `_Can_memmove_in_copy` form, as it clearly separates the CRT function from the STL algorithm with `_in_` (the problem with `move` and `copy` is that they are so short and so generic-sounding that they don't immediately parse as STL algorithm names when inside larger identifiers; `_in_` seems to help a lot because it makes them clearly nouns). I'd be fine with `_Can_safely_CRTFUNCTION_in_STLFUNCTION` too, which could be applied to `_Equal_memcmp_is_safe` and `_Fill_memset_is_safe`.

miscco commented:
That sounds reasonable, especially given that I always run around requesting better names

miscco commented:
@AdamBucior I have changed the names to `_Can_memop_in_algorithm` and also moved the `_Equal_memcmp_is_safe` and `_Fill_memset_is_safe` helper ofer to the new name.

Are you ok with this or should we move this to `_Memop_in_algorithm_is_safe`. (This is slightly longer so ...)

AdamBucior commented:
> @AdamBucior I have changed the names to `_Can_memop_in_algorithm` and also moved the `_Equal_memcmp_is_safe` and `_Fill_memset_is_safe` helper ofer to the new name.
> 
> Are you ok with this or should we move this to `_Memop_in_algorithm_is_safe`. (This is slightly longer so ...)

I'm not a maintainer so it's not for me to decide.

miscco commented:
> I'm not a maintainer so it's not for me to decide.

You are a contributor, so you can certainly voice your preference. I mean I do not have a strong preference, so If you say `_Memop_in_algorithm_is_safe` is way better I am fine renaming them.

AdamBucior commented:
> > I'm not a maintainer so it's not for me to decide.
> 
> You are a contributor, so you can certainly voice your preference. I mean I do not have a strong preference, so If you say `_Memop_in_algorithm_is_safe` is way better I am fine renaming them.

I personally prefer `_Memop_in_algorithm_is_safe`.

miscco commented:
@AdamBuior, moved to `_Memop_in_algorithm_is_safe`

CaseyCarter commented:
/azp run

miscco commented:
I would like to add a rename here too:
`_Copy_memmove`                  ==> `_Memmove_forward`
`_Copy_memmove_backward` ==> `_Memmove_backward`
`_Copy_memmove_common`  ==> `_Memmove_forward_bounded`

We use these functions all over the place in `copy` and `move` so the `_Copy` prefix just adds confusion as to what is happening.

Any disagreement

AdamBucior commented on stl/inc/xutility:
```diff
@@ -4313,72 +4313,116 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+        bool_constant<_Can_memop_elements<remove_cv_t<_Source>, remove_volatile_t<_Dest>, _Cat>>>;
+
+template <class _IterSource, class _IterDest, _Memop_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memop<move_iterator<_IterSource>, _IterDest, _Cat> =
```
Shouldn't `_Copy` and `_Copy_uninitialized` become `_Move` and `_Move_uninitialized` with `move_iterator`? (I know it's pre-existing but it's definitely wrong)

miscco replied:
I think that you are right that for the `_Copy` case it seems wrong. For `_Copy_uninitialized` we require `trivial` types anyway where it does not make a difference.

Thatsaid I am absolutely in favor of changing it accordingly

AdamBucior replied:
Since we fixed it in #1772 it's no longer pre-existing so we should change this accordingly.

StephanTLavavej reviewed (changes requested):
This PR has merge conflicts now that we're using `if constexpr` unconditionally.

miscco commented:
@CaseyCarter I have rebased and updated the PR. I think we should decide whether this is worth it or simply put it in the bin

miscco commented on stl/inc/memory:
```diff
@@ -78,7 +78,8 @@ namespace ranges {
             _STL_INTERNAL_STATIC_ASSERT(constructible_from<iter_value_t<_Out>, iter_reference_t<_It>>);
 
-            if constexpr (is_same_v<_Se, _It> && is_same_v<_OSe, _Out> && _Ptr_copy_cat<_It, _Out>::_Really_trivial) {
+            if constexpr (is_same_v<_Se,
+                              _It> && is_same_v<_OSe, _Out> && _Memmove_in_uninitialized_copy_is_safe<_It, _Out>) {
```
The gods have failed me :(

StephanTLavavej reviewed (changes requested):
There are a number of merge conflicts - can you resolve them?

miscco commented:
Yeah, will do over the weekend

StephanTLavavej reviewed (changes requested):
After merging your #1771, there's a conflict in `<vector>`.

miscco commented:
So I am going to drop this and replace it with a sequence of smaller less error prone PRs:

1. Add variable templates for the type aliases and use them in the Codebase
2.-N Refactor the implementation of each variable template into a right thing.

Does that sound like a plan?

AdamBucior commented:
Yeah, I think it's a good idea to split this into smaller PRs.
